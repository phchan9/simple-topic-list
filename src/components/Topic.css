/* make ordered-list decimal-indexed */
.Topic {
    list-style: decimal inside;
    display: list-item;
    border: 0 none;
}
