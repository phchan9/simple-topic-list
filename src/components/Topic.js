import React from 'react';
import PropTypes from 'prop-types';

// make ordered-list decimal-indexed
const topicStyle = {
  listStyle: 'decimal inside',
  display: 'list-item',
  border: '0 none',
};

/* eslint-disable react/jsx-filename-extension */
const Topic = ({ title, upVotes, downVotes, onUpVoteCallBack, onDownVoteCallBack }) => (
  <li style={topicStyle} className="list-group-item">
    <h5>{title}</h5>
    <div>
      <button
        type="button"
        className="btn btn-default"
        onClick={onUpVoteCallBack}
      >
        UpVote <span className="badge">{upVotes}</span>
      </button>
      <button
        type="button"
        className="btn btn-default"
        onClick={onDownVoteCallBack}
      >
        DownVote <span className="badge">{downVotes}</span>
      </button>
    </div>
  </li>
);

Topic.propTypes = {
  title: PropTypes.string.isRequired,
  upVotes: PropTypes.number.isRequired,
  downVotes: PropTypes.number.isRequired,
  onUpVoteCallBack: PropTypes.func.isRequired,
  onDownVoteCallBack: PropTypes.func.isRequired,
};

export default Topic;
