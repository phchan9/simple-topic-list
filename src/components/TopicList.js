import React from 'react';
import PropTypes from 'prop-types';
import Topic from './Topic';

const TopicList = ({ topics, onUpVoteCallBack, onDownVoteCallBack}) => (
  <div>
    <ol className="list-group">
      {topics.map(topic => (
        <Topic
          key={topic.id}
          onUpVoteCallBack={() => onUpVoteCallBack(topic.id, topic.upVotes + 1)}
          onDownVoteCallBack={() => onDownVoteCallBack(topic.id, topic.downVotes + 1)}
          {...topic}
          />
      ))}
    </ol>
  </div>
);

TopicList.propTypes = {
  topics: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    upVotes: PropTypes.number.isRequired,
    downVotes: PropTypes.number.isRequired
  })).isRequired,
  onUpVoteCallBack: PropTypes.func.isRequired,
  onDownVoteCallBack: PropTypes.func.isRequired
};

export default TopicList;
