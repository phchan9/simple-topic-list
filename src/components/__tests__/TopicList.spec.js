import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TopicList from '../TopicList';
import Topic from '../Topic';
import createTopic from '../../utils';

/* eslint-disable react/jsx-filename-extension */
const setup = () => {
  const props = {
    topics: [
      createTopic(1, 23, 34, 'title1'),
      createTopic(2, 32, 43, 'title2'),
    ],
    onUpVoteCallBack: jest.fn(),
    onDownVoteCallBack: jest.fn(),
    fetchTopics: jest.fn(),
  };

  const wrapper = shallow(<TopicList {...props} />);

  return {
    props,
    wrapper,
  };
};

describe('TopicList Component', () => {
  it('render correctly', () => {
    const { wrapper } = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should properly fire `onUpVoteCallBack` or `onDownVoteCallBack` or `fetchTopics`when click any of child components',
    () => {
      const { wrapper, props } = setup();
      wrapper.find(Topic).forEach((component) => {
        component.prop('onUpVoteCallBack')();
        expect(props.onUpVoteCallBack).toHaveBeenCalled();

        component.prop('onDownVoteCallBack')();
        expect(props.onDownVoteCallBack).toHaveBeenCalled();
      });
    });
});
