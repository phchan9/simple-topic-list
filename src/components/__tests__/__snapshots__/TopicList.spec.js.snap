// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicList Component render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <ol
        className="list-group"
    >
        <Topic
            downVotes={34}
            id={1}
            onDownVoteCallBack={[Function]}
            onUpVoteCallBack={[Function]}
            title="title1"
            upVotes={23}
        />
        <Topic
            downVotes={43}
            id={2}
            onDownVoteCallBack={[Function]}
            onUpVoteCallBack={[Function]}
            title="title2"
            upVotes={32}
        />
    </ol>
</div>,
  "nodes": Array [
    <div>
      <ol
            className="list-group"
      >
            <Topic
                  downVotes={34}
                  id={1}
                  onDownVoteCallBack={[Function]}
                  onUpVoteCallBack={[Function]}
                  title="title1"
                  upVotes={23}
            />
            <Topic
                  downVotes={43}
                  id={2}
                  onDownVoteCallBack={[Function]}
                  onUpVoteCallBack={[Function]}
                  title="title2"
                  upVotes={32}
            />
      </ol>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <TopicList
        onDownVoteCallBack={[Function]}
        onUpVoteCallBack={[Function]}
        topics={
                Array [
                        Object {
                          "downVotes": 34,
                          "id": 1,
                          "title": "title1",
                          "upVotes": 23,
                        },
                        Object {
                          "downVotes": 43,
                          "id": 2,
                          "title": "title2",
                          "upVotes": 32,
                        },
                      ]
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "onDownVoteCallBack": [Function],
          "onUpVoteCallBack": [Function],
          "topics": Array [
            Object {
              "downVotes": 34,
              "id": 1,
              "title": "title1",
              "upVotes": 23,
            },
            Object {
              "downVotes": 43,
              "id": 2,
              "title": "title2",
              "upVotes": 32,
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <ol
                    className="list-group"
          >
                    <Topic
                              downVotes={34}
                              id={1}
                              onDownVoteCallBack={[Function]}
                              onUpVoteCallBack={[Function]}
                              title="title1"
                              upVotes={23}
                    />
                    <Topic
                              downVotes={43}
                              id={2}
                              onDownVoteCallBack={[Function]}
                              onUpVoteCallBack={[Function]}
                              title="title2"
                              upVotes={32}
                    />
          </ol>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <ol
                    className="list-group"
          >
                    <Topic
                              downVotes={34}
                              id={1}
                              onDownVoteCallBack={[Function]}
                              onUpVoteCallBack={[Function]}
                              title="title1"
                              upVotes={23}
                    />
                    <Topic
                              downVotes={43}
                              id={2}
                              onDownVoteCallBack={[Function]}
                              onUpVoteCallBack={[Function]}
                              title="title2"
                              upVotes={32}
                    />
          </ol>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <TopicList
    onDownVoteCallBack={[Function]}
    onUpVoteCallBack={[Function]}
    topics={
        Array [
            Object {
              "downVotes": 34,
              "id": 1,
              "title": "title1",
              "upVotes": 23,
            },
            Object {
              "downVotes": 43,
              "id": 2,
              "title": "title2",
              "upVotes": 32,
            },
          ]
    }
/>,
}
`;
